### PR TITLE
Tag picker fixes

### PR DESCRIPTION
**Pull request by Reettaphant**
out
- Apologies for the long PR: by the time I realised I needed to rearrange the tag picker for this, I had already started on many of the improvements. 

- This PR makes many improvements to the usability of the tag pickers
- It hides tag search results when user is no longer focuses on the input
- It allows for navigating up and down search results using arrow keys
- It fixes padding issues in the byline picker
- It changes the commissioning and keyword tag picker presentation to be more in line with what is on composer (showing selected tag as a vertical list, not on the input element itself)
- It also breaks the tag picker into multiple components to make it more usable and to allow for more customisation.

### CODE REVIEW

File: public/video-ui/src/components/FormFields/TagPicker.js [hunk 1]
```diff
@@ -34,359 +43,264 @@ export default class TagPicker extends React.Component {
...
-      })
-    ) {
-      return {
-        id: this.state.inputString,
-        webTitle: this.state.inputString
-      };
-    }
-    return [];
+  removeFn = (tag) => {
+    const newFieldValue = this.state.tagValue.filter(oldField => {
```
Comment (akash1810) [inline public/video-ui/src/components/FormFields/TagPicker.js]: `filter` returns a list, does this work as expected?

File: public/video-ui/src/components/FormFields/TagPicker.js [hunk 2]
```diff
@@ -1,19 +1,28 @@
...
+
   state = {
     capiTags: [],
-    inputString: '',
-    lastAction: UserActions.other,
     tagValue: [],
-    capiUnavailable: false
+    capiUnavailable: false,
+    showTags: true,
+    tagsVisible: false,
```
Comment (akash1810) [inline public/video-ui/src/components/FormFields/TagPicker.js]: it looks like `showTags` and `tagsVisible` are related and can be merged into a single property

File: public/video-ui/src/components/FormFields/TagPicker.js [hunk 3]
```diff
@@ -34,359 +43,264 @@ export default class TagPicker extends React.Component {
...
-  renderTags(tag) {
-    const addTag = () => {
-      const valueWithoutStringDupes = removeStringTagDuplicates(
-        tag,
-        this.state.tagValue
-      );
-      const newFieldValue = valueWithoutStringDupes.concat([tag]);
+    if (e.keyCode === keyCodes.enter && this.state.selectedTagIndex !== null) {
+      const newTag = this.state.capiTags[this.state.selectedTagIndex];
+      const valueWithoutDupes = this.props.tagType === TagTypes.contributor ? removeStringTagDuplicates(newTag, this.state.tagValue) : removeTagDuplicates(newTag, this.state.tagValue);
```
Comment (akash1810) [inline public/video-ui/src/components/FormFields/TagPicker.js]: the linter would cry about this line length 😜

File: public/video-ui/src/components/FormFields/TagPicker.js [hunk 4]
```diff
@@ -34,359 +43,264 @@ export default class TagPicker extends React.Component {
...
+      const newFieldValue = valueWithoutDupes.concat([newTag]);
 
       this.setState({
-        inputString: ''
+        selectedTagIndex: null,
+        inputClearCount: this.state.inputClearCount + 1
```
Comment (akash1810) [inline public/video-ui/src/components/FormFields/TagPicker.js]: what's `inputClearCount` used for?

File: public/video-ui/src/components/FormFields/TagPicker.js [hunk 5]
```diff
@@ -34,359 +43,264 @@ export default class TagPicker extends React.Component {
...
-  renderCapiUnavailable() {
-    if (this.state.capiUnavailable) {
-      return (
-        <span className="form__field--external__error">
-          Tags are currently unavailable
-        </span>
-      );
-    }
+      newTagValues.push(tagValue);
+      return newTagValues;
```
Comment (akash1810) [inline public/video-ui/src/components/FormFields/TagPicker.js]: minor: could use a [spread operator](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Operators/Spread_operator) here:

```js
return [...newTagValues, tagValue];
```

File: public/video-ui/src/components/FormFields/TagPicker.js [hunk 6]
```diff
@@ -34,359 +43,264 @@ export default class TagPicker extends React.Component {
...
+    }
   }
 
-  renderValue = (field, i) => {
-    const removeFn = () => {
-      const newFieldValue = this.state.tagValue.filter(oldField => {
-        return field.id !== oldField.id;
-      });
+  onSort = (sortedList) => {
+    const newTagValues = sortedList.reduce((newTagValues, sortedValue) => {
```
Comment (akash1810) [inline public/video-ui/src/components/FormFields/TagPicker.js]: could we use a distinct name for the accumulator for clarity?

File: public/video-ui/src/components/FormFields/TagPicker.js (hunk 2 above)
Comment (Reettaphant) [inline public/video-ui/src/components/FormFields/TagPicker.js]: Do you mean two different properties in the same object?

File: public/video-ui/src/components/FormFields/TagPicker.js (hunk 1 above)
Comment (Reettaphant) [inline public/video-ui/src/components/FormFields/TagPicker.js]: Yes, `onUpdate` converts the list into a string before saving the value but the tagpicker handles the tags as lists.

File: public/video-ui/src/components/FormFields/TagPicker.js (hunk 4 above)
Comment (Reettaphant) [inline public/video-ui/src/components/FormFields/TagPicker.js]: To send a message from the `TagPicker` to its children to clear the input string because a new tag has been selected by navigating the tag list with arrow keys. If the count changes, the children know they should set the input string to an empty string.

File: public/video-ui/src/components/CapiSearch/CapiSearch.js [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+      }
+    }
+  }
+
+  renderTags(tag, index) {
+
+    const getTagClassName = () => {
+      const name = "form__field__tags" + (index === this.props.selectedTagIndex ? " form__field__tags--selected" : "");
+
+      return "form__field__tags" + (index === this.props.selectedTagIndex ? " form__field__tags--selected" : "");
```
Comment (Reettaphant) [inline public/video-ui/src/components/CapiSearch/CapiSearch.js]: I like this pattern. It saves a lot of typing. This would only be worth doing I think it were abstracted out to a utility function which adds a new classname to the end of the string if a condition is satisfied. But maybe this would be another pr?

File: public/video-ui/src/components/CapiSearch/CapiUnavailable.js [hunk 8]
```diff
@@ -0,0 +1,17 @@
+ import React from 'react';
+
+ export default class CapiUnavailable extends React.Component {
+
+  render() {
+    if (this.props.capiUnavailable) {
+      return (
+        <div className="form__field--external__error">
```
Comment (Reettaphant) [inline public/video-ui/src/components/CapiSearch/CapiUnavailable.js]: Yes, it doesn't make any sense. I'll fix it

File: public/video-ui/src/components/FormFields/TagPicker.js (hunk 2 above)
Comment (akash1810) [inline public/video-ui/src/components/FormFields/TagPicker.js]: from what I understood from the code, `showTags` is always the inverse of `tagsVisible` which made me think it could be one boolean value?

Comment (Reettaphant) [inline public/video-ui/src/components/FormFields/TagPicker.js]: It's not. `tagsVisible` is used to check if `onBlu` event was triggered by clicking on one of the pickers children (this should not hide the tags) or elsewhere (this should hide the tags). `showTags` is set to true/false depending on whether or not we should display the tag list. I should probably rename that variable because it is not very descriptive.
